lib/rust/ast/src/generation.rs

```diff
@@ -0,0 +1,324 @@
+    fn extends(&mut self, ident:&Ident) {
+        if let Some(name) = self.extends.get(&ident).cloned() {
+            write!(self.code, " extends ");
+            self.typ_name(&name);
+        }
+        writeln!(self.code, "");
+    }
+
+    /// Generates scala type parameters.
+    ///
+    /// `<Foo, Bar>` = `[Foo, Bar]`
+    fn generics(&mut self, generics:&syn::Generics) {
+        if generics.params.is_empty() {return}
+        write!(self.code, "[");
+        for (i, param) in generics.params.iter().enumerate() {
+            if i != 0 { write!(self.code, ", "); }
+            match param {
+                syn::GenericParam::Type(typ) => self.typ_name(&typ.ident),
+                _ => (),
+            }
+        }
+        write!(self.code, "]");
+    }
+
+    /// Generates a qualified scala type with type arguments.
+    ///
+    /// `foo::Bar<Baz>` => `Foo.Bar[Baz]`
+    fn typ(&mut self, typ:&syn::Type) {
+        match typ {
+            syn::Type::Path(path) => {
+                for (i, typ) in path.path.segments.iter().enumerate() {
+                    if i != 0 { write!(self.code, "."); }
+                    self.typ_segment(typ);
+                }
+            }
+            _ => (),
+        }
+    }
+
+    /// Generates a scala type with type arguments.
+    ///
+    /// `Foo<Bar<Baz>>` => `Foo[Bar[Baz]]`
+    fn typ_segment(&mut self, typ:&syn::PathSegment) {
+        let boxed = typ.ident.to_string().as_str() == "Box";
+        if !boxed { self.typ_name(&typ.ident); }
+        match &typ.arguments {
+            syn::PathArguments::AngleBracketed(typ) => {
+                if !boxed { write!(self.code, "["); }
+                for (i, typ) in typ.args.iter().enumerate() {
+                    if i != 0 { write!(self.code, ", "); }
+                    match typ {
+                        syn::GenericArgument::Type(typ) => self.typ(typ),
+                        _ => (),
+                    }
+                }
+                if !boxed { write!(self.code, "]"); }
+            }
+            _ => (),
+        }
+    }
+
+    /// Generates a scala variable name (camel case).
+    ///
+    /// `foo_bar` => `fooBar`
+    fn var_name(&mut self, ident:&Ident) {
+        let mut underscore = false;
+        for char in ident.to_string().chars() {
+            if char == '_' {
+                underscore = true ;
+            } else if underscore {
+                underscore = false;
+                for char in char.to_uppercase() {
+                    self.code.push(char)
+                }
+            } else {
+                self.code.push(char);
+            }
+        }
+    }
+
+    /// Generates a scala type name.
+    ///
+    /// Converts rust primitives into scala primitives i.e. `usize` => `Int`
+    fn typ_name(&mut self, ident:&Ident) {
+        let name = match ident.to_string().as_str() {
+            "usize" | "isize" | "u64" | "u32" | "u16" | "i64" | "i32" | "i16" | "i8" => "Int",
```

Comment:
This isn't true. Anything that's 64-bit defined by the Rust standard should map to `Long`. `Int` on the JVM is fixed at 32-bit. Something like `usize`can also convert to `Long` as we assume that we run on 64-bit platforms only. This asumption should be documented, however.